tap_dynamodb/deserialize.py

```diff
@@ -58,23 +58,61 @@ def _apply_projection(self, record, breadcrumb, output):
                 breadcrumb_key = breadcrumb[0].split('[')[0]
                 index = int(breadcrumb[0].split('[')[1].split(']')[0])
                 if output.get(breadcrumb_key):
-                    output[breadcrumb_key].append(record[breadcrumb_key][index])
+                    # record: {'Artist': 'No One You Know5', 'metadata': ['test1']}
+                    # main breadcrumb = [['metadata[0]'], ['metadata[1]']]
+                    # current breadcrumb = ['metadata[1]']
+                    # current output = {'metadata': ['test1']}
+                    # expected output = {'metadata': ['test1']}
+                    #       as "metadata" has only 1 item and the current breadcrumb is expecting 2nd item
+
+                    # only prepare output if the list field contains data at that index position in record
+                    if len(record.get(breadcrumb_key)) > index:
+                        output[breadcrumb_key].append(record[breadcrumb_key][index])
                 else:
-                    output[breadcrumb_key] = [record[breadcrumb_key][index]]
+                    # record: {'Artist': 'No One You Know5'}
+                    # main breadcrumb = [['metadata[0]']]
+                    # current breadcrumb = ['metadata[0]']
+                    # current output = {}
+                    # expected output = {'metadata': []}
+                    #       as "metadata" does not have any items and the current breadcrumb is expecting 1st item
 
+                    output[breadcrumb_key] = []
+                    # only prepare output if the list field contains data at that index position in record
+                    if record.get(breadcrumb_key) and len(record.get(breadcrumb_key)) > index:
+                        output[breadcrumb_key].append(record[breadcrumb_key][index])
             else:
                 output[breadcrumb[0]] = record.get(breadcrumb[0])
         else:
             if '[' in breadcrumb[0]:
                 breadcrumb_key = breadcrumb[0].split('[')[0]
                 index = int(breadcrumb[0].split('[')[1].split(']')[0])
-                if output.get(breadcrumb_key) is None:
+                if not output.get(breadcrumb_key):
                     output[breadcrumb_key] = [{}]
-                self._apply_projection(record[breadcrumb_key][index], breadcrumb[1:], output[breadcrumb_key][0])
+
+                # record: {'Artist': 'No One You Know5'}
+                # main breadcrumb = [['metadata[0]', 'Age']]
+                # current breadcrumb = ['metadata[0]', 'Age']
+                # current output = {'metadata': [{}]}
+                # expected output = {'metadata': [{}]}
+                #       as "metadata" is not present and the current breadcrumb is expecting 1st item and which is a parent
+
+                # only prepare output if the list field contains data at that index position in record
+                if record.get(breadcrumb_key) and len(record.get(breadcrumb_key)) > index:
```

Comment:
Same here where this should be 
```suggestion
                if breadcrumb_key in record and len(record.get(breadcrumb_key)) > index:
```

Reply:
Updated.